flutter_appauth_platform_interface/lib/src/errors.dart

```diff
@@ -0,0 +1,112 @@
+import 'package:flutter/services.dart';
+
+/// The details of an error thrown from the underlying
+/// platform AppAuth libraries.
+class FlutterAppAuthPlatformErrorDetails {
+
+  FlutterAppAuthPlatformErrorDetails({
+    required this.type,
+    required this.code,
+    required this.error,
+    required this.errorDescription,
+    required this.errorUri,
+    required this.rootCauseError,
+    required this.userDidCancel,
+  });
+
+  /// On iOS: One of the domain values here: 
+  ///         https://github.com/openid/AppAuth-iOS/blob/c89ed571ae140f8eb1142735e6e23d7bb8c34cb2/Sources/AppAuthCore/OIDError.m#L31
+  /// On Android: One of the type codes here: 
+  ///         https://github.com/openid/AppAuth-Android/blob/c6137b7db306d9c097c0d5763f3fb944cd0122d2/library/java/net/openid/appauth/AuthorizationException.java
+  /// Recommend not using this field unless you really have to, see `error` field below.
+  final String? type;
+
+  /// On iOS: One of the enum values defined here depending on the error type
+  ///         https://github.com/openid/AppAuth-iOS/blob/c89ed571ae140f8eb1142735e6e23d7bb8c34cb2/Sources/AppAuthCore/OIDError.h
+  /// On Android: One of the codes defined here:
+  ///          https://github.com/openid/AppAuth-Android/blob/c6137b7db306d9c097c0d5763f3fb944cd0122d2/library/java/net/openid/appauth/AuthorizationException.java#L158
+  /// Recommend not using this field unless you really have to, see `error` field below.
+  final String? code;
+
+  /// For 400 errors from the Authorization server, this is error defined here: https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+  /// e.g "invalid_grant"
+  /// Otherwise a short error describing what happened.
+  final String? error;
+
+  /// Short, human readible error description.
+  final String? errorDescription;
+
+  /// The error Uri or domain from the underlying platform.
+  final String? errorUri;
+
+  /// The underlying raw error as a String debugging.
+  final String? rootCauseError;
```

Comment:
A few points on this one
- I don't understand what this means so I suspect others won't either
- The description doesn't line up with the name so something seems off here
- I took a look further on this and on iOS this references a `debugDescription`. Apple's own [docs](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418703-debugdescription) say this is textual representation of the object itself for debugging purposes. This reads like the equivalent of calling `toString()` on the object itself and aligns with what I've seen the values I've seen.  On Android, this references a cause and this means the original error/exception. Based on this, these refer to different things so wouldn't be appropriate to map to the same property. I would suggest keeping this to Android only and mention as much in the API docs. Unless I'm missing something, the `debugDescription` from iOS seems redundant given it's a single string representation that consists of the code, domain and user info. Code and domain are already returned. If the user info dictionary is useful then I would suggest having a separate property that is documented as being specific to iOS/macOS

Reply:
I slightly disagree with this, though ultimately it's your call. (I've updated the comment though since it's misleading).

The underlying error (the `cause` field on Android and the error on iOS) are untyped properties so theoretically could contain other bits of information that I've missed. Especially in future changes to the underlying platform's AppAuth SDKs which might be easy to miss while upgrading this library.

Having a catch all property that ensures everything is exposed to dart code means that developers are always guaranteed to have all the available information in some form, if nothing else it makes it easier to then raise a PR on this library to add additional properties.

Reply:
I have a suspicion what I said has been misunderstood so will try to elaborate. Based on your response, I would guess that you think I'm saying to not have this property or to remove information. If so, that isn't what I'm saying. I understand your intent to have a catch all property and have no issue with providing more information. My main point is what this represents on each platform is different. For iOS/macOS, this represents the error itself. It's only on Android that this represents what **caused** of the error. This means it represents another error entirely that the name of `rootCauseError` only make sense for Android. It also explains why on Android that when you close the browser that the `rootCauseError` is an empty string. This is because it doesn't contain the details of the `AuthorizationException` itself, which would've made it consistent with iOS/macOS. Rather, it contains the details of another error/exception that caused the `AuthorizationException` to be thrown.

Hope this is clearer and why I mentioned having two separate properties (one for iOS and one for Android) makes more sense.

Reply:
Oops sorry looks like I made a mistake myself. Just realised again that on iOS you do look at the underlying error. Sorry for the confusion. Let me mark this as resolved as I spotted something else and will comment on the relevant part(s) of the PR